server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -612,12 +612,46 @@ public static class Builder {
 
             private boolean hasSyntheticSourceCompatibleKeywordField;
 
+            /**
+             * Returns whether the given builder should be used to track source for synthetic source purposes.
+             */
+            private boolean shouldTrackSource(KeywordFieldMapper.Builder keywordBuilder) {
+                // we only need one compatible keyword multi-field, so if one was already found, ignore the rest
+                return hasSyntheticSourceCompatibleKeywordField == false && isSyntheticSourceCompatible(keywordBuilder);
+            }
+
+            /**
+             * Returns whether the given builder is synthetic source compatible. To be compatible, the builder must
+             * track source in some way, whether that be via the "store" param or using doc values.
+             */
+            private boolean isSyntheticSourceCompatible(KeywordFieldMapper.Builder keywordBuilder) {
```

Comment:
Do you think we can make `MultiFields.Builder` an interface and do this stuff in a decorator or something inside `TextFieldMapper`? I know we have `hasSyntheticSourceCompatibleKeywordField` already but this is growing more and more.